R/generate_correlated_data.R

```diff
@@ -586,3 +586,115 @@ genCorOrdCat <- function(dtName, idname = "id", adjVar = NULL, baseprobs,
+#' @examples
+#' genBlockMat(rho = 0.6, nInds = 2, nPeriods = 3, corstr = "ind")
+#' genBlockMat(rho = 0.6, nInds = 2, nPeriods = 3, corstr = "cs")
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "ar1")
+#' 
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "cs", iRho = 0.8)
+#' genBlockMat(rho = 0.6, nInds = 3, nPeriods = 3, corstr = "ar1", iRho = 0.8)
+#' @export
+#' @concept correlated
+genBlockMat <- function(rho, nInds, nPeriods, corstr = "ind", 
+                        iRho = NULL) {
+  
+  ### Checking
+  
+  assertNotMissing(rho = missing(rho), nInds = missing(nInds),
+                   nPeriods = missing(nPeriods))
+  
+  assertInteger(nInds = nInds, nPeriods = nPeriods)
+  assertAtLeast(nPeriods = nPeriods, minVal = 2)
+  assertInRange(rho = rho, range = c(-1,1))
+  
+  if (!is.null(iRho)) {
+    assertInRange(iRho = iRho, range = c(-1,1))
+    if (length(rho) == 1) {
+      assertNotEqual(corstr = corstr, val = "ind", 
+        msg = " With iRho set, corstr should be either cs or ar1.")  
+    }
+  }
+  
+  if (length(rho) > 1) {
+    assertLength(rho = rho, length = nPeriods)
+    if (!is.null(iRho)) {
+      assertLength(iRho = iRho, length = (nPeriods - 1))
+    }
+  }
+  
+  assertOption(corstr = corstr, options = c("ind", "cs", "ar1"))
+  
+  ###
+  
+  assignDiag <- function(block, value) {
+    diag(block) <- value
+    return(block)
+  }
+  
+  ### create rvec from rho
+  
+  if (length(rho) == nPeriods) {
+    rvec <- rho
+  } else {
+    if (corstr == "ind") {
+      rvec <- vapply(1:nPeriods, function(x) rho*(x==1), numeric(1))
+    } else if (corstr == "cs") {
+      rvec <- vapply(1:nPeriods, function(x) rho, numeric(1))
+    } else if (corstr == "ar1") {
+      rvec <- vapply(1:nPeriods, function(x) rho^x, numeric(1))
+    }
+  }
+  
+  ### create irvec from iRho
+  
+  if (! is.null(iRho)) {
+    
+    if (length(iRho) < (nPeriods - 1)) {
+      if (corstr == "cs") {
+        irvec <- vapply(2:nPeriods, function(x) iRho, numeric(1))  
+      } else if (corstr == "ar1") {
+        irvec <- vapply(2:nPeriods, function(x) iRho^(x-1), numeric(1))
+      }
+    } else { irvec <- iRho }
+    irvec <- c(1, irvec)
+  } else if (is.null(iRho)) {
+    irvec <- c(1, rvec[2:nPeriods])
+  }
+  
+  ### generate blocks
+  
+  blocks <- lapply(1:nPeriods, function(x) matrix(rvec[x], nInds, nInds))
+  blocks <- lapply(1:nPeriods, function(x) assignDiag(blocks[[x]], irvec[x]))
+  
+  diag(blocks[[1]]) <- 1
+  names(blocks) <- paste0("M", c(1 : nPeriods))
+  
+  ### put blocks together
+  
+  block_str <- sapply(nPeriods:1, 
+                      function(x) data.table::shift(1:nPeriods, nPeriods - x, fill = 0))
+  block_str[upper.tri(block_str)] = t(block_str)[upper.tri(block_str)]
+  block_str <- matrix(paste0("M", block_str), nPeriods)
+  
+  bm <- blockmatrix::blockmatrix(value = block_str, list = blocks, dim=c(nPeriods, nPeriods))
```

Comment:
As {blockmatrix} is only a suggests you should guard this function with `requireNamespace` see: https://r-pkgs.org/dependencies.html#guarding-the-use-of-a-suggested-package

Reply:
Done - thanks.